LocalRateLimit(HTTP): Add dynamic token bucket support

vikaschoudhary16 opened the pull request:
<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message: LocalRateLimit(HTTP): Add dynamic token bucket support
Additional Description:
fixes: https://github.com/envoyproxy/envoy/issues/23351 and https://github.com/envoyproxy/envoy/issues/19895

User configures [descriptors](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/common/ratelimit/v3/ratelimit.proto#envoy-v3-api-msg-extensions-common-ratelimit-v3-localratelimitdescriptor) in the [http local rate limit filter](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto). These descriptors are the "target" to match using the source descriptors built using the traffic(http requests). Only matched traffic will be rate limited. When request comes, at runtime, based on [rate_limit](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-msg-config-route-v3-ratelimit) configuration, descriptors are generated where `values` are picked from the request as directed by the rate_limit configuration. These generated descriptors are matched with "target"(user configured) descriptors. Generated descriptors are very flexible already in the sense that "values" from the request can be extracted using number of ways such as dynamic metadata, matcher api, computed reg expressions etc etc, but in "target"(user configured) descriptors are very rigid and it is expected that user statically configures the "values" in the descriptor. 
This PR is adding flexibility by allowing blank "values" in the user configured descriptors. Blank values will be treated as wildcard. Suppose descriptor entry key is `client-id` and value is left blank by the user, local rate limit filter will create a descriptor dynamically for each unique value of header `client-id`. That means `client1`, `client2` and so on will have dedicated descriptors and token buckets.

To keep a resource consumption under limit, LRU cache is maintained for dynamic descriptors with a default size of 20, which is configurable.


Docs Changes: TODO
Release Notes: TODO

wbpcode commented:
I think we still need a way to limit the overhead and memory of the token buckets. It's unacceptable to let it increases unlimited.

vikaschoudhary16 commented:
> I think we still need a way to limit the overhead and memory of the token buckets. It's unacceptable to let it increases unlimited.

Thanks a lot for taking a look.
Will add logic to keep it under limits

wbpcode commented:
Thanks for this contribution. Dynamic descriptor support is a very complex problem in the local rate limit, considering various limitations. 

Before flush more comments to the implementation details, I will throw some quesions first:
1. Should the LRU list be updated when one of the entry is hit in the worker? I think it should to keep the active descriptor alive. But this will also introduces more compleixity.
2. Should different descriptor has independent list? Or the descriptor with much more different dynamic values will occupies the whole list?
3. Lifetime management and search logic is much more complex now.

So, I will suggest you to take a step back and think is there other way to resolve your requirement, like a new limit filter in your fork or using the rate_limit/rate_limit_quota directly.

If you insist on to enhance the local_rate_limit, then, I think we should provide an abstraction (like DynamicDescriptor) to wrap all these complexity first (list/memory management, lifetime problem, cross workers updating, etc.), and them, we could integrate the high level abstraction into the local_rate_limit. Or it's hard for reviewer to ensure the quality of the exist feature.

Thanks again for all your help and contribution. This is not simple feature. 🌷

vikaschoudhary16 commented:
> Thanks for this contribution. Dynamic descriptor support is a very complex problem in the local rate limit, considering various limitations.
> 
> I have take a pass to current implementation, but before flush more comments to the implementation details, I will throw some quesions first:
> 
> 1. Should the LRU list be updated when one of the entry is hit in the worker? I think it should to keep the active descriptor alive. But this will also introduces more compleixity.
> 2. Should different descriptor has independent list? Or the descriptor with much more different dynamic values will occupies the whole list?
> 3. Lifetime management and search logic is much more complex now.
> 
> So, I will suggest you to take a step back and think is there other way to resolve your requirement, like a new limit filter in your fork or using the rate_limit/rate_limit_quota directly.
> 
> If you insist on to enhance the local_rate_limit, then, I think we should provide an abstraction (like DynamicDescriptor) to wrap all these complexity first (list/memory management, lifetime problem, cross workers updating, etc. (PS: I think lock may be an option if we can limit the lock only be used in the new feature)), and then, we could integrate the high level abstraction into the local_rate_limit. Or it's hard for reviewer to ensure the quality of the exist feature.
> 
> Thanks again for all your help and contribution. This is not simple feature. 🌷


Thanks a lot @wbpcode for taking a look. Really appreciate!!
> So, I will suggest you to take a step back and think is there other way to resolve your requirement, like a new limit filter in your fork or using the rate_limit/rate_limit_quota directly.

From the conversations on the linked issues, seems like there has been repetitive request for this functionality since long time, so figuring out a solution in community might benefit number of users.
yeah, rate_limit_quota can address the usecase in case user is fine with external dependency. In some cases, users are keen in avoiding external dependency and looking for solution through local rate limiter path.

> If you insist on to enhance the local_rate_limit, then, I think we should provide an abstraction (like DynamicDescriptor) to wrap all these complexity first (list/memory management, lifetime problem, cross workers updating, etc. (PS: I think lock may be an option if we can limit the lock only be used in the new feature)), and then, we could integrate the high level abstraction into the local_rate_limit. 

Sounds good. I will work on adding the abstraction as per your suggestion.

wbpcode commented:
Thanks so much for this update. I think this make sense. Here are some high level suggestions to this (I think we are in the correct way, thanks):

1. Please add an explict bool flag to enable this feature to avoid some users mis-use this when they configure an empty descriptor in case.
2. This is a new feature. So, you can ignore the timer based token bucket completely. It would make the implementation simper.
3. You may need to refactor the return type of `requestAllowed`. Now, because the buckets in the dynamic descriptor may be destroyed in another thread.

wbpcode commented on api/envoy/extensions/filters/http/local_ratelimit/v3/local_rate_limit.proto:
```diff
@@ -167,4 +167,13 @@ message LocalRateLimit {
+  // Actual number of dynamic descriptors will depend on the cardinality of unique values received from the http request for the omitted
+  // values.
+  // Default is 20.
+  uint32 dynamic_descripters_lru_cache_limit = 18;
```
May name this simplely `max_dynamic_descripters`. (We may change the elimination algorithm in future, who know?) and please use wrapper number type `google.protobuf.UInt32Value`.

And Please add explict bool to enable this feature, like `google.protobuf.BoolValue use_dynamic_descripters`.

vikaschoudhary16 replied:
done

vikaschoudhary16 commented:
> Thanks so much for this update. I think this make sense. Here are some high level suggestions to this (I think we are in the correct way, thanks):
> 
> 1. Please add an explict bool flag to enable this feature to avoid some users mis-use this when they configure an empty descriptor in case.
> 2. This is a new feature. So, you can ignore the timer based token bucket completely. It would make the implementation simper.
> 3. You may need to refactor the return type of `requestAllowed`. Now, because the buckets in the dynamic descriptor may be destroyed in another thread.

Hey @wbpcode , apologies for a very long delay in updating the PR. I got trapped in another downstream activity and then holidays came.

Have addressed your comments. Will be working on ci and test coverage. Meanwhile, when you get chance can you please take a look.  _/\_

duxin40 commented:
hello! sorry to interrupt you, i meet a CI problem for PreCheck as below, and i found that your ci also meet this problem~
Do you have any thoughts on this problem? I think we can communicate. I think it can be solved by re-building the bazel remote cache, but I don’t know how to trigger the rebuild of the cache.
![image](https://github.com/user-attachments/assets/e31f2384-ffa0-4c17-ae02-7170ef058b93)

here is my PR: https://github.com/envoyproxy/envoy/pull/36667

vikaschoudhary16 commented:
> hello! sorry to interrupt you, i meet a CI problem for PreCheck as below, and i found that your ci also meet this problem~ Do you have any thoughts on this problem? I think we can communicate. I think it can be solved by re-building the bazel remote cache, but I don’t know how to trigger the rebuild of the cache. ![image](https://private-user-images.githubusercontent.com/33946910/404746604-e31f2384-ffa0-4c17-ae02-7170ef058b93.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzczNDk5MzUsIm5iZiI6MTczNzM0OTYzNSwicGF0aCI6Ii8zMzk0NjkxMC80MDQ3NDY2MDQtZTMxZjIzODQtZmZhMC00YzE3LWFlMDItNzE3MGVmMDU4YjkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIwVDA1MDcxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlOWMzZmU4N2MxNDVkZjI1OGUzOTI4YWU1YTRhZDk0N2Y4MGY5YTk3YmM3NTFlZDkzOTJiMDM4YjIzZjM4OGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.L3q8QifoMiJOASms3Ti8bYQU1UyjXDV0HOBlcbhVo1g)
> 
> here is my PR: #36667

I dont know how to do that. may be @phlax can help/guide us to handle it better

I just pushed an empty commit to re-trigger ci and it dint happen next time.
`git commit  -s --allow-empty`

duxin40 commented:
> > hello! sorry to interrupt you, i meet a CI problem for PreCheck as below, and i found that your ci also meet this problem~ Do you have any thoughts on this problem? I think we can communicate. I think it can be solved by re-building the bazel remote cache, but I don’t know how to trigger the rebuild of the cache. ![image](https://private-user-images.githubusercontent.com/33946910/404746604-e31f2384-ffa0-4c17-ae02-7170ef058b93.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzczNDk5MzUsIm5iZiI6MTczNzM0OTYzNSwicGF0aCI6Ii8zMzk0NjkxMC80MDQ3NDY2MDQtZTMxZjIzODQtZmZhMC00YzE3LWFlMDItNzE3MGVmMDU4YjkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTIwVDA1MDcxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRlOWMzZmU4N2MxNDVkZjI1OGUzOTI4YWU1YTRhZDk0N2Y4MGY5YTk3YmM3NTFlZDkzOTJiMDM4YjIzZjM4OGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.L3q8QifoMiJOASms3Ti8bYQU1UyjXDV0HOBlcbhVo1g)
> > here is my PR: #36667
> 
> I dont know how to do that. may be @phlax can help/guide us to handle it better
> 
> I just pushed an empty commit to re-trigger ci and it dint happen next time. `git commit -s --allow-empty`

+1 ! @phlax could you please help us? If you can help us it will be very much appreciated!

The reason I can think of currently is because my PR was proposed in October 24, and before merging the latest main code a few days ago, CI could run normally, but after merging the latest main code (my other The code has basically not been changed), and the PreCheck part of CI started to report errors.
The following is the commit that I merged into the latest main code:https://github.com/envoyproxy/envoy/commit/914297e6c69f4c4a3fe313d5496a628bf795efc5#diff-3cba4d66ffa4e29abbcf8cb94c7489e8451af43fdad6859d4bbcfd17058419e2

vikaschoudhary16 commented:
same whisperer error again on ci.

duxin40 commented:
> the problem is likely to resolve itself (once the relevant cache gets blown) - will keep monitoring

thanks @phlax  for helping us！okay i will wait for some tjme until the cache refresh～

adisuissa commented:
@wbpcode PTAL.

wbpcode commented:
Thanks for this great contribution. I am so happy to see this happens. I think we are at correct way after a quick check. But I prefer to review and land this after https://github.com/envoyproxy/envoy/pull/38197 because it will also simplify this PR.

vikaschoudhary16 commented:
> Thanks for this great contribution. I am so happy to see this happens. I think we are at correct way after a quick check. But I prefer to review and land this after #38197 because it will also simplify this PR.

sure, I am anyways not assuming timer token bucket in this PR as per your previous suggestion so will not impact much.

alyssawilk commented:
/wait on the other PR.  please main merge when it's ready for another pass, and it'll show back up in the oncall dashboard

wbpcode commented on source/common/runtime/runtime_features.cc:
```diff
@@ -162,6 +162,9 @@ FALSE_RUNTIME_GUARD(envoy_reloadable_features_google_grpc_disable_tls_13);
 // before downstream.
 FALSE_RUNTIME_GUARD(envoy_reloadable_features_allow_multiplexed_upstream_half_close);
 
+// A flag to enable the usage of dynamic buckets for local rate limiting. For example, dynamically
+// created token buckets for each unique value of a request header.
+FALSE_RUNTIME_GUARD(envoy_reloadable_features_local_rate_limiting_with_dynamic_buckets);
```
I think runtime flag is unnecessary because dynamic descriptors is only enabled when the empty descriptor `value` is set which is not allowed in the past.

I we can treat this is a new feature that controlled by explict API configuration.

wbpcode commented on source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc:
```diff
@@ -158,55 +183,162 @@ LocalRateLimiterImpl::Result LocalRateLimiterImpl::requestAllowed(
+
+// Compare the request descriptor entries with the user descriptor entries. If all non-empty user
+// descriptor values match the request descriptor values, return true
+bool DynamicDescriptorMap::matchDescriptorEntries(
+    const std::vector<RateLimit::DescriptorEntry>& request_entries,
+    const std::vector<RateLimit::DescriptorEntry>& user_entries) {
```
The user_entries is a little unclear. It would be better to call it `config_entries` or something.

wbpcode commented on source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc:
```diff
@@ -158,55 +183,162 @@ LocalRateLimiterImpl::Result LocalRateLimiterImpl::requestAllowed(
+    return false;
+  }
+
+  bool has_empty_value = false;
+  for (size_t i = 0; i < request_entries.size(); ++i) {
+    // Check if the keys are equal
+    if (request_entries[i].key_ != user_entries[i].key_) {
+      return false;
+    }
+
+    // all non-blank user values must match the request values
+    if (!user_entries[i].value_.empty() && user_entries[i].value_ != request_entries[i].value_) {
+      return false;
+    }
+
+    // Check for empty value in user entries
+    if (user_entries[i].value_.empty()) {
+      has_empty_value = true;
+    }
+  }
+  return has_empty_value;
```
```suggestion
  for (size_t i = 0; i < request_entries.size(); ++i) {
    // Check if the keys are equal.
    if (request_entries[i].key_ != user_entries[i].key_) {
      return false;
    }
    
    // Check values are equal or wildcard value is used.
    if (user_entries[i].value_.empty()) {
      continue;
    }
    if (request_entries[i].value_ != user_entries[i].value_) {
      return false;
    }
  }
  return true;
```

wbpcode reviewed (commented):
Thanks for the update. Some comments are added. And please check all the comments to ensure it match our style. Thanks again!

vikaschoudhary16 commented on source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc:
```diff
@@ -109,8 +111,26 @@ LocalRateLimiterImpl::LocalRateLimiterImpl(
+          throw EnvoyException("minimum allowed value for max_dynamic_descriptors is 1");
+        }
+
+        if (per_connection) {
+          throw EnvoyException(
+              "local rate descriptor value cannot be empty in per connection rate limit mode");
+        }
```
not necessary. Just wanted to reduce scope and cover it in follow up. I think it will require just passing `max_dynamic_descriptors` from filter config to `PerConnectionRateLimiter`.

vikaschoudhary16 commented on source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.h:
```diff
@@ -84,14 +122,15 @@ class RateLimitTokenBucket : public TokenBucketContext {
 using RateLimitTokenBucketSharedPtr = std::shared_ptr<RateLimitTokenBucket>;
 
-class LocalRateLimiterImpl {
+class LocalRateLimiterImpl : public Logger::Loggable<Logger::Id::rate_limit_quota> {
```
added a new id `local_rate_limit`

wbpcode replied:
I think you forget to change this line code?

vikaschoudhary16 commented on source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc:
```diff
@@ -109,8 +111,26 @@ LocalRateLimiterImpl::LocalRateLimiterImpl(
+                               "descriptor value cannot be empty");
+        }
+
+        if (lru_size == 0) {
+          throw EnvoyException("minimum allowed value for max_dynamic_descriptors is 1");
+        }
```
added PGV to cover this

wbpcode commented on source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc:
```diff
@@ -158,55 +170,159 @@ LocalRateLimiterImpl::Result LocalRateLimiterImpl::requestAllowed(
+RateLimitTokenBucketSharedPtr
+DynamicDescriptorMap::getBucket(const RateLimit::Descriptor request_descriptor) {
+  for (const auto& pair : user_descriptors_) {
+    auto user_descriptor = pair.first;
```
nit: config_descriptor

wbpcode commented on source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc:
```diff
@@ -158,55 +170,159 @@ LocalRateLimiterImpl::Result LocalRateLimiterImpl::requestAllowed(
+      continue;
+    }
+
+    // we found a user configured wildcard descriptor that matches the request descriptor.
```
Please all check all these comments match our style.

vikaschoudhary16 replied:
I have updated this specific comment. Is there a doc/readme that have guidelines about the comment styles for my reference?

wbpcode reviewed (commented):
LGTM with very very minor comments. Thanks for update 🙏

wbpcode reviewed (approved):
LGTM. Thanks.
